### ISSUE TITLE
bool8, bool16 Not Blittable

### ISSUE THREAD

**Issue by chadefranklin** [MrUnbelievable92/MaxMath#10]
When attempting to send bool8 and/or bool16 to a static burst compiled method, I get a burst error that these types are not blittable due to the "asArray" field. Removing the field did fix the issue. Are you aware of this? Is this desired?

**Comment by MrUnbelievable92** [comment]
Oof.
No, i was not aware of it. It was a hack for a more efficient array accessor operator, before Burst released the `Unity.Burst.CompilerServices.Constant.IsConstantExpression<T>` compiler intrinsic, the usage i.e. implementation of which is in my backlog.
Gonna work on that and create a smaller release soon™.

Thanks so much! I don't use function pointers so I wouldn't know.

**Comment by chadefranklin** [comment]
Awesome! It's probably the case for bool32 as well though I haven't tested.

Somewhat related, I am curious if & why all the conversions to and from byte are necessary. If there is a better place to talk about that could we do so there?

**Comment by MrUnbelievable92** [comment]
We can do it here, "publically" - I don't really mind.

It is not necessary but in Burst-ed code at least it is free in release mode (like... completely omitted).
The main reason for it being that way is avoidance of code duplication due to all the benefits it entails. Secondly this throws exceptions in DEBUG if any of your bools are neither 0 nor 1 as bytes, which could lead to serious issues, since I wrote the native SIMD code in the a way that expects the data to be either of those values.
It's probably bad for managed C# unless you're building with IL2CPP which uses clang a.k.a. LLVM aswell (i .e. the same compiler Burst targets), so it will be omitted there, too.

It's an old relic by now and I can't see a reason as to why I shouldn't just copy/paste the "direct" way of doing all that stuff so yeah - I'll do that, too. Same goes for all `quarter` vectors (the 8 bit float type) and `float8`.

**Comment by chadefranklin** [comment]
Thanks for the explanation. I did also want to pass some other off-topic ideas, findings, and questions by you though (there is a bit more than mentioned here. It would be nice to have a faster communication channel such as discord, but it's fine if not). Please let me know if there is reasoning against any suggestions as I am fairly new to SIMD programming.

I am experimenting with some wider types (int16 and float16) and was having some trouble with those byte conversions not being omitted as you say (due to my incorrect use i believe).

I did notice (at least one instance of) your SIMD code that "expects the data to be either of those values" and rewrote it to follow the Unity method of doing it for bool4 which doesn't have that requirement:
`mm256_cmpeq_epi32( mm256_cvtepi8_epi32(bool8), v256(0) )`

Another thing I've been experimenting with is mask types. I see you mentioned something similar in your latest release notes: "reduced latency of boolean operations when used with an expression as the argument". What do you think about "float8 < float8" returning a "mask256" type -_which has an implicit conversion to bool8_- instead of returning bool8 directly? Then certain operations such as Select statements can use the mask256 instead of bool8 bypassing the conversion to/from. This seems to work well in my testing so far. It seems this is how things are done by default outside of Unity though _my_ experience with SIMD is limited to the Unity environment.

**Comment by MrUnbelievable92** [comment]
Most of the things that oppose specific implementations that aim for encapsulation of implementation details of the same thing are not necessarily related to a particular ISA, but rather what the compiler you target can reason about. It's a constant battle with the compiler.
In this case, it is also about the reverse operation of `cmpeq(ZERO, boolVec)`. Converting a mask of n-bit -1s and 0s to 8-bit 1s and 0s has to be be done in a very specific way so that the compiler still knows what you're doing, especially when you have to convert it back to its original state in the software layer when - for example - using `select` with an expression as the bool vector. Those go hand in hand, which is most likely why your experiments were not successful; I've also been through that but I believe that I have found the only way of doing that after about 2 years of experimentation :D BUT it is the by far (we're talking a handful of clock cycles here and bytes of code size, of course, but still) the less efficient method, though, when writing a `bool` vector back to memory or some specific functions like `any`/`all` that use `tzcnt` under the hood or even reading out a vector element in the source code, and it's the same with Unitys or rather LLVMs approach. Sometimes there just is an inherent trade-off between different approaches where optimized, hand written assembly or intrinsics code is simply unmatched.
The only issue I face, currently, is when you cross the 256bit to 128bit boundary. Unitys `double3` and `double4` DO actually perform this optimally but at this point I believe that there is some "magic" going on where Burst tells LLVM that these types map directly to vector types it knows about, which was also mentioned in a talk by a Burst compiler engineer.

The "mask256" type is interesting and I haven't though about that, mostly because the goal of this kind of library is abstracting away all those details, and for me it is too technical to be a part of the public API. The only issue I see right now is that a "mask256" has to have implicit type conversions to many different `bool` types, those being all `bool` vector types from `bool3` onward. That's actually a good starting point, though, because it can be essentially hidden away completely because of implicit type conversions - I will log it and experiment with it.

Regarding the faster channel of communication: I find that there is a lot of value in having such things being discussed and archived here in the issues tab so that any third party can also give their opinions on things they possibly know more about than either of us. I really don't mind you possibly spamming a ton of minor issues - on the contrary. Nevertheless I have set up a rudimentary discord channel [here](https://discord.gg/KvWpmnWK); we can definitely have a talk. If you want to, please let's agree on a specific date and time due to possible time zone differences (UTC+1 here) and/or work schedules.

**Comment by chadefranklin** [comment]
Any chance you remember the name of that talk you mention?

I agree the mask types are a bit technical, and yeah, I didn't think about all the conversions that would be necessary. The mask types may be a potential solution to that "magic" going on though. I do believe my experiments have been successful though (from looking at the burst inspector & getting expected results and performance improvements) in my limited use case (float16, int16, bool4x4-mask512).

**Comment by MrUnbelievable92** [comment]
Just want to archive that we looked at your solutions and they don't work either.

The talk is [this one](https://www.youtube.com/watch?v=BpwvXkoFcp8&t=1369s). The "magic" of linking Unity.Mathematics types to LLVM types is mentioned twice there.

**Comment by chadefranklin** [comment]
Thanks, I thought it might be that one. It was a great talk.

Only with the mask types have I been able to eliminate the vector-bool unnecessary conversion. I am currently using custom comparison extension methods (hidden behind a sub-namespace) to generate the masks. For example, `float8.gt(float8)` will generate a mask256 from a greater than check. I then added a select statement overload to accept mask256. I would be ok with this workflow as it is completely "opt-in" for more performance-hungry users (me) to squeeze in that extra bit.
